Add a new AddLink() operation to Span. (#3678)
open-telemetry/opentelemetry-specification: specification/trace/api.md

# Tracing API

> **Status**: [Stable, Feature-freeze](../document-status.md), except where otherwise specified

Reviewer: Formally, wouldn't we have to lift the feature freeze before adding a new feature? I think feature freeze was only meant for limited time to close releases, maybe it's only there by mistake?

Author: (Yes, I think it is a mistake)

[wrapping it into a Span](#wrapping-a-spancontext-in-a-span).
For example, a `Propagator` performing context extraction may need this.

> #### Specifying links

Reviewer 2: Why is this requirement removed? I think implementations should still be required to allow users to add links during creation, right?

#### Specifying links

During `Span` creation, a user MUST have the ability to record links to other `Span`s.
Linked `Span`s can be from the same or a different trace -- see [links](#link).
`Link`s added at `Span` creation may be considered by [Samplers](sdk.md#sampler)

> to make a sampling decision

Reviewer 3: ```suggestion
to make a sampling decision.
```

Note that [`RecordException`](#record-exception) is a specialized variant of
`AddEvent` for recording exception events.

#### Add Link

> 

Reviewer 4: Mark this section as Experimental?

> A `Span` MUST have the ability to add `Link`s associated with it after its creation - see [Links](#link).

Reviewer 4: Which one do we want?
1. Link can be added while IsRecording=true https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#isrecording.
2. Link can be added at any time after the span is created (regardless of where IsRecording is true or false).

Reviewer 5: Describing the shape of the workflow is an activity that is independent of whether those activities are being recorded or not.

Reviewer 6: @reyang I think your question bridges into other issues. See 
https://github.com/open-telemetry/opentelemetry-specification/issues/2918
and
https://github.com/open-telemetry/opentelemetry-specification/issues/2986.

Independent of the changes here, I think we should consider investing in sampler API changes. This has been discussed in Sampling SIG and we are blocked on the need for prototypes. Here are some statements on this topic that are relevant here:

- We need a way to record spans because their links are sampled, despite the span context itself not being sampled, call this the IS-EXPORTING quality as distinct from the IS-RECORDING quality
- Work in the Sampling SIG ([ongoing](https://github.com/open-telemetry/oteps/pull/235 ), [prior](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-probability-sampling.md)) gives us a way to indicate when a span was exported but not sampled (a.k.a. "zero adjusted count")
- I do not believe do not believe it is feasible to change the is-recording state of a span it starts
- Jaeger supports a way to defer a sampling decision until the first child span starts, which means some AddLinks could be considered before the sampling decision
- It is sensible (see #2918, #2986) to allow a change of IS-EXPORTING after the start event, meaning to allow a span to decide to export because (1) IS-RECORDING was set at start, and (2) something interesting happened (e.g.,  AddLink event occurred from a sampled context).

None of this is meant to block this PR, these are proposed sampling improvements, where prototypes are welcome. (If this topic interests anyone, please join the Sampling SIG weekly on Thursdays, see the OTel public calendar.)

Author: @reyang Is this resolved (having as follow up the issues mentioned by jmacd)?

> 
> Observe that `Link`s added after `Span` creation are not considered by [Samplers](sdk.md#sampler)
> to make a sampling decision.

Reviewer 5: ```suggestion
`Link`s added after `Span` creation may not be considered by [Samplers](sdk.md#sampler)
```
* no need to break paragraphs for single sentence
* "Observe that" is unnecessary 
* we don't know that links are not considered by samplers, this is not part of the spec. A different implementation might consider them.

| `CONSUMER` | | yes | maybe | |
| `INTERNAL` | | | | |

## Link

A user MUST have the ability to record links to other `Span`s.
Linked `Span`s can be from the same or a different trace -- see [Links
between spans](../overview.md#links-between-spans).

A `Link` is structurally defined by the following properties:

- `SpanContext` of the `Span` to link to.
- Zero or more [`Attributes`](../common/README.md#attribute) further describing
  the link.

The API MUST provide:

- An API to record a single `Link` where the `Link` properties are passed as
  arguments. This MAY be called `AddLink`. This API takes the `SpanContext` of
  the `Span` to link to and optional `Attributes`, either as individual
  parameters or as an immutable object encapsulating them, whichever is most
  appropriate for the language. Implementations MAY ignore links with an

>   [invalid](#isvalid) `SpanContext`.

Reviewer 2: This seems odd. I think it was copied from above, but as it is structured this seems to be talking about a top-level API.

Reviewer 6: Yes. @carlosalberto I suggest restoring the former "#### Specifying links" section above with less text now, saying APIs MUST support specifying one or more links with just the from 426-428 (except the last sentence):

```
During `Span` creation, a user MUST have the ability to record links to other
`Span`s. Linked `Span`s can be from the same or a different trace -- see [Links
between spans](../overview.md#links-between-spans).
```

This section keeps
```
A `Link` is structurally defined by the following properties:

- `SpanContext` of the `Span` to link to.
- Zero or more [`Attributes`](../common/README.md#attribute) further describing
  the link.

Span SHOULD preserve the order in which `Link`s are set.

Note that [Samplers](sdk.md#sampler) may only consider information already
present during span creation. Any `Link`s added later may not change the sampling
decision.
```

and what remains belongs in the AddLink section. Thanks :grin:

Author: @jmacd Feedback (partially applied) ;)

@MrAlias I don't read it that way - now we have both (back) "Specifying links" and "AddLink" operations at Span, which point to this section (we have the same for `SpanKind` for example).

| `CONSUMER` | | yes | maybe | |
| `INTERNAL` | | | | |

> ## Link

Reviewer: I am missing the SDK perspective. Will this call be implemented using some conventions on top of AddEvent, or is there an OTLP extension expected?

Linked `Span`s can be from the same or a different trace -- see [Links
between spans](../overview.md#links-between-spans).

A `Link` is structurally defined by the following properties:

- `SpanContext` of the `Span` to link to.
- Zero or more [`Attributes`](../common/README.md#attribute) further describing
  the link.
- `timestamp`, default to current time. This argument SHOULD only be set
  when span creation time has already passed. If API is called at a moment of
  a Span logical start, API user MUST NOT explicitly set this argument.

The Span creation API MUST provide:

- An API to record a single `Link` where the `Link` properties are passed as
  arguments. This MAY be called `AddLink`. This API takes the `SpanContext` of
  the `Span` to link to and optional `Attributes` and `timestamp`, either as individual
  parameters or as an immutable object encapsulating them, whichever is most
  appropriate for the language. Implementations MAY ignore links with an
  [invalid](#isvalid) `SpanContext`.

The Span interface MAY provide:

- An API to add multiple `Link`s at once, where the `Link`s are passed in a
  single method call.

> Links SHOULD preserve the order in which they're set.

Reviewer 5: I think you mean "Spans should preserve the order"

> Note that [Samplers](sdk.md#sampler) can only consider information already

Reviewer 5: incorrect - depends on SDK/Sampler implementation

| `CONSUMER` | | yes | maybe | |
| `INTERNAL` | | | | |

## Link

> A user MUST have the ability to record links to other `Span`s.

Reviewer 7: nit: 
```suggestion
A user MUST have the ability to record links to other `SpanContext`s.
```

> Linked `Span`s can be from the same or a different trace -- see [Links

Reviewer 7: ```suggestion
Linked `SpanContext`s can be from the same or a different trace -- see [Links
```

between spans](../overview.md#links-between-spans).

A `Link` is structurally defined by the following properties:

- `SpanContext` of the `Span` to link to.
- Zero or more [`Attributes`](../common/README.md#attribute) further describing
  the link.

The API MUST provide:

- An API to record a single `Link` where the `Link` properties are passed as
  arguments. This MAY be called `AddLink`. This API takes the `SpanContext` of
  the `Span` to link to and optional `Attributes`, either as individual
  parameters or as an immutable object encapsulating them, whichever is most
  appropriate for the language. Implementations MAY ignore links with an
  [invalid](#isvalid) `SpanContext`.

The Span interface MAY provide:

- An API to add multiple `Link`s at once, where the `Link`s are passed in a
  single method call.

Span SHOULD preserve the order in which `Link`s are set.

> Note that [Samplers](sdk.md#sampler) may only consider information already
> present during span creation. Any `Link`s added later may not change the sampling
> decision.
> 

Reviewer 3: ```suggestion
The API documentation MUST state that adding links at span creation is preferred to calling `AddLink` later, as samplers can only consider information already present during span creation.
```

This wording was requested by some in a previous attempt to add this feature.

@jmacd Given that our mental model around sampling and links has evolved since then, do you still think it would be valuable adding this?

Reviewer 6: How does this sound?
```suggestion
The API documentation MUST state that adding links at span creation is preferred to calling `AddLink` later, for contexts that are available during span creation, because head sampling decisions can only consider information present during span creation.
```
